File: src/main/java/de/tectoast/emolga/commands/various/NicknameCommand.java
package de.tectoast.emolga.commands.various;

import de.tectoast.emolga.bot.EmolgaMain;
import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.CommandCategory;
import de.tectoast.emolga.commands.GuildCommandEvent;
import de.tectoast.emolga.utils.Constants;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.TextChannel;
import org.jsolf.JSONObject;

public class NicknameCommand extends Command {
    public NicknameCommand() {
        super("nickname", "Ändert deinen Nickname (funktioniert nur 1x pro Woche)", CommandCategory.Various, Constants.BSID, Constants.ASLID);
        setArgumentTemplate(ArgumentManagerTemplate.builder().add("name", "Neuer Name", "Der neue Nickname", ArgumentManagerTemplate.Text.any())
                .setExample("!nickname IchMagEmolga")
                .build());
    }

    @Override
    public void process(GuildCommandEvent e) {
        TextChannel tco = e.getChannel();
        Message m = e.getMessage();
        String msg = m.getContentDisplay();
        Member member = e.getMember();
        String nickname = msg.substring(10);
        if (!e.getGuild().getSelfMember().canInteract(member)) {
            tco.sendMessage("Ich kann deinen Nickname nicht modifizieren!").queue();
            return;
        }
        if (nickname.length() > 32) {
            tco.sendMessage("Dieser Nickname ist zu lang! (Er darf maximal 32 Zeichen enthalten)").queue();
            return;
        }
        try {
            JSONObject json = getEmolgaJSON();
            Guild g = tco.getGuild();
            if (!json.has("cooldowns")) json.put("cooldowns", new JSONObject());
            if (!json.getJSONObject("cooldowns").has(g.getId()))
                json.getJSONObject("cooldowns").put(g.getId(), new JSONObject());
            if (json.getJSONObject("cooldowns").getJSONObject(g.getId()).has(member.getId())) {
                long l = Long.parseLong(json.getJSONObject("cooldowns").getJSONObject(g.getId()).getString(member.getId()));
                long untilnow = System.currentTimeMillis() - l;
                if (untilnow < 604800000) {
                    long delay = 604800000 - untilnow;
                    int days = (int) (delay / 86400000);
                    delay -= days * 86400000L;
                    int hours = (int) (delay / 3600000);
                    delay -= hours * 3600000L;
                    int minutes = (int) (delay / 60000);
                    tco.sendMessage(member.getAsMention() + " Du kannst deinen Namen noch nicht wieder ändern!\nCooldown: " + days + "d " + hours + "h " + minutes + "m").queue();
                    return;
                }
            }
            String oldname = member.getEffectiveName();
            member.modifyNickname(nickname).complete();
            if (g.getId().equals("518008523653775366"))
                EmolgaMain.emolgajda.getGuildById("518008523653775366").getTextChannelById("728675253924003870").sendMessage(oldname + " hat sich in " + nickname + " umbenannt!").queue();
            tco.sendMessage(member.getAsMention() + " Dein Nickname wurde erfolgreich geändert!").queue();
            json.getJSONObject("cooldowns").getJSONObject(g.getId()).put(member.getId(), Long.toString(System.currentTimeMillis()));
            saveEmolgaJSON();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/various/NicknameCommand.kt
package de.tectoast.emolga.commands.various

import com.mongodb.client.model.Filters.and
import com.mongodb.client.model.ReplaceOptions
import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.CommandCategory
import de.tectoast.emolga.commands.GuildCommandEvent
import de.tectoast.emolga.utils.Constants
import de.tectoast.emolga.utils.json.Cooldown
import de.tectoast.emolga.utils.json.db
import de.tectoast.emolga.utils.json.eq
import de.tectoast.emolga.utils.json.findOne
import dev.minn.jda.ktx.coroutines.await

class NicknameCommand : Command(
    "nickname",
    "Ändert deinen Nickname (funktioniert nur 1x pro Woche)",
    CommandCategory.Various,
    Constants.G.BS,
    Constants.G.ASL
) {
    init {
        argumentTemplate = ArgumentManagerTemplate.builder()
            .add("name", "Neuer Name", "Der neue Nickname", ArgumentManagerTemplate.Text.any())
            .setExample("!nickname IchMagEmolga").build()
        slash(true, Constants.G.ASL)
    }

    override suspend fun process(e: GuildCommandEvent) {
        val member = e.member
        val nickname = e.arguments.getText("name")
        if (!e.guild.selfMember.canInteract(member)) {
            e.reply("Ich kann deinen Nickname nicht modifizieren!")
            return
        }
        if (nickname.length > 32) {
            e.reply("Dieser Nickname ist zu lang! (Er darf maximal 32 Zeichen enthalten)")
            return
        }
        val g = e.guild

        //secondsToTime()
        val gid = g.idLong
        val uid = member.idLong
        db.cooldowns.findOne(and(Cooldown::guild eq gid, Cooldown::user eq uid))?.run {
            val expiresIn = this.timestamp - System.currentTimeMillis()
            if (expiresIn <= 0) return@run
            e.reply(
                "${member.asMention} Du kannst deinen Namen noch nicht wieder ändern!\nCooldown: ${
                    secondsToTime(expiresIn / 1000)
                }"
            )
            return
        }
        //val oldname = member.effectiveName
        member.modifyNickname(nickname).await()
        /*if (g.idLong == Constants.G.ASL) g.getTextChannelById("728675253924003870")!!
            .sendMessage("$oldname hat sich in $nickname umbenannt!").queue()*/
        e.reply(member.asMention + " Dein Nickname wurde erfolgreich geändert!")
        db.cooldowns.replaceOne(
            and(Cooldown::guild eq gid, Cooldown::user eq uid),
            Cooldown(gid, uid, System.currentTimeMillis() + 604800000),
            ReplaceOptions().upsert(true)
        )
    }
}
